### DISCUSSION TITLE
Destructuring query tuple types shown as {unknown} in Rust Analyzer

### DISCUSSION THREAD

**Discussion by yngwi** [bevyengine/bevy#5184] [Q&A] [answered]
Hi,

I'm starting to learn Rust and Bevy using Neovim with Rust Analyzer as language server. So far it works just fine. I have one question, though:

When I have the following function:

```rust
fn greet_people(query: Query<(&Name, &Position), (With<Person>, With<Position>)>) {
    for (name, pos) in query.iter() {
        println!("hello {} at {}/{}!", name.0, pos.x, pos.y);
    }
}
```

Rust analyzer shows the types of `name` and `pos` as `{unknown}`. Using this kind of destructuring on other things like `let (x, y) = ("x", "y");` shows the correct types for `x` and `y`. Is there something I can do to also get the correct type hints for the query iterator?

Thanks,
Daniel

**Comment by bjorn3** [comment, marked as answer]
Either chalk is not able to resolve all types or you don't have proc macros enabled.

**Reply by yngwi** [reply]
Hi. Thank you for your reply. I think I have it enabled, this is my language server configuration:
```lua
settings = {
	["rust-analyzer"] = {
		checkOnSave = {
			features = "all",
			command = "clippy",
		},
		procMacro = {
			enable = true,
		},
	},
},
```

I am seeing the following error message in the LSP Logs:

`[ERROR][2022-07-03 17:22:55] .../vim/lsp/rpc.lua:430	"rpc"	"rust-analyzer"	"stderr"	"thread 'MacroExpander' panicked at 'assertion failed: `(left != right)`\n  left: `0`,\n right: `0`', crates\\proc-macro-srv\\src\\abis\\abi_1_63\\proc_macro\\bridge\\handle.rs:22"
`

Unfortunately I have no Idea what that means or if it is indeed related to the problem.

**Reply by yngwi** [reply]
https://github.com/rust-lang/rust-analyzer/issues/12600 This issue seems to be about this kind of crash, maybe it is indeed related...

**Reply by bjorn3** [reply]
That is probably the cause. You can switch to stable or beta rustc until we add support for the new proc macro abi the latest nightly versions use.

**Reply by yngwi** [reply]
ok thank you
